core-v1/src/main/scala/engine/SimpleCachingEngine.scala

```diff
@@ -210,6 +241,27 @@ object SimpleCachingEngine {
       debugging(expr).invokeAndReturn(state(i, cached(o)))
     }
 
+    override def visitContainsAny[I, W[+_] : Extract, C[_] : Foldable, A, B : OP](
+      expr: Expr.ContainsAny[I, W, C, A, B, OP],
+    ): I => CachedResult[B] = memoize(expr, _) { i =>
+      val CachedResult(input, inputCacheState) = expr.inputExpr.visit(this)(i)
+      val CachedResult(wrappedValidSet, finalCacheState) =
+        visitWithUpdatedCache(i, expr.validValuesExpr, inputCacheState)
+      // TODO: Implement short-circuiting
+      // TODO: Move to CommonEngine
+      val W = Extract[W]
+      val valid = wrappedValidSet.map(W.extract)
+      val found = input
+        .foldRight(Eval.now(List.empty[W[A]])) { (wa, acc) =>
+          val a = W.extract(wa)
+          if (valid(a)) acc.map(wa :: _)
+          else acc
+        }
+        .value
+      val o = expr.foundMatchingElements(input, wrappedValidSet, found)
+      debugging(expr).invokeAndReturn(state((i, input, wrappedValidSet, found), cached(o, finalCacheState)))
+    }
```

Comment:
These TODOs will be impacted by the refactor of `Expr.ContainsAny`, so I am going to postpone the work on these until those changes are made.